### PR TITLE
add rename fields and hydration input for complex path

### PR DESCRIPTION
**Pull request by andimarek**

### CODE REVIEW

File: src/main/java/graphql/nadel/engine/transformation/CollapseTransformation.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        // skip traversing subtree because the fields are in respect to the underlying schema and not the overall which will break
+        return TraversalControl.ABORT;
+    }
+
+
+    @Override
+    public ExecutionResultNode unapplyResultNode(ExecutionResultNode executionResultNode, List<FieldTransformation> transformations, UnapplyEnvironment environment) {
+        LeafExecutionResultNode leafExecutionResultNode = getLeafNode(executionResultNode);
+        // path and type is still wrong here
+        return changeFieldInResultNode(leafExecutionResultNode, getOriginalField());
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/transformation/CollapseTransformation.java]: is it ok that this is still wrong??  Is this a comment that you meant to come back to?

File: src/main/java/graphql/nadel/engine/transformation/FieldUtils.java [hunk 2]
```diff
@@ -15,4 +23,27 @@ public static String resultKeyForField(Field field) {
...
+            if (curField != null) {
+                newField.selectionSet(newSelectionSet().selection(curField).build());
+            }
+            newField.name(path.get(ix));
+            curField = newField.build();
+        }
+        return curField;
+    }
+
+    public static LeafExecutionResultNode getLeafNode(ExecutionResultNode executionResultNode) {
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/transformation/FieldUtils.java]: getFirstLeafNode would be a better name and some JavaDoc saying that would be even better!

Comment (andimarek) [reply src/main/java/graphql/nadel/engine/transformation/FieldUtils.java]: renamed

File: src/main/java/graphql/nadel/NadelTuple2.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
+package graphql.nadel;
+
+import graphql.Internal;
+
+import java.util.Objects;
+
+
+@Internal
+public class NadelTuple2<T1, T2> {
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/NadelTuple2.java]: Just call it Tuple2 - thats what package names are for

File: src/main/java/graphql/nadel/engine/UnapplyEnvironment.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    public TraverserContext<ExecutionResultNode> context;
+    public ExecutionStepInfo parentExecutionStepInfo;
+    public boolean isHydrationTransformation;
+    public boolean batched;
+    public Map<String, String> typeRenameMappings;
+    public GraphQLSchema overallSchema;
+    public List<Field> notTransformedFields;
+
```
Comment (bbakerman) [inline src/main/java/graphql/nadel/engine/UnapplyEnvironment.java]: public getters please.  I know its painful.

Wats to address this

* allow a new dependency like Immutables or AutoValve
* use an IDEA plugin like BOB
* writee out own IDEA plugin

**Pull request merged**